### ISSUE TITLE
SKIPREST

### ISSUE THREAD

**Issue by loveduah1** [OPM/opm-simulators#3846]
```
   61 Reading WELSPECS in PROJ1.DATA line 364
   62 Reading COMPDAT  in PROJ1.DATA line 374
   63 Reading WCONPROD in PROJ1.DATA line 385
   64 Reading WCONINJE in PROJ1.DATA line 394
   65 Reading TSTEP    in PROJ1.DATA line 405

Warning: Unsupported keywords or keyword items:

  NOECHO: keyword not supported
  In file: PROJ1.DATA, line 75
  Will continue

  ECHO: keyword not supported
  In file: PROJ1.DATA, line 107
  Will continue

Creating grid from keywords DX, DY, DZ and TOPS

Error: Problem with keyword RESTART
In PROJ1.DATA line 255
The restart file: PROJ1.X0002 does not exist

Error: Unrecoverable errors were encountered while loading input
```
[SKIPREST.zip](https://github.com/OPM/opm-simulators/files/8327453/SKIPREST.zip)

Please SKIPREST keyword returns the error above. attached is the file I ran.

**Comment by bska** [comment]
> Please `SKIPREST` keyword returns the error above

In this case it's not actually `SKIPREST`.  Rather it's the `RESTART` keyword:

> ```
> Error: Problem with keyword RESTART
> In PROJ1.DATA line 255
> The restart file: PROJ1.X0002 does not exist
> ```

The `RESTART` keyword in your input file is
```
SOLUTION
-- -------------------------------------------------------------------------
--item #:      1        2         3          4    
RESTART
            PROJ1      2         1*         1*  /
```
Moreover, you've got `UNIFOUT` in your `RUNSPEC` section which means that the base run produces unified output files (`PROJ1.SMSPEC`, `PROJ1.UNSMRY`, and `PROJ.UNRST`).

Did you already produce a base run that you're trying to restart from?  If not, the `RESTART` keyword is not appropriate here.  If you **did** produce a base run, do you have the `PROJ1.UNRST` file somewhere?  If you do, then please insert the keyword `UNIFIN` next to the `UNIFOUT` keyword in the `RUNSPEC` section.  You will also use a more complete path if the `PROJ1.UNRST` file is in a different location than your `PROJ1.DATA` file.

Otherwise, if you have separate restart files named `PROJ1.X*` (e.g., `PROJ1.X0002`) somewhere that's not in the same directory as your `PROJ1.DATA`, then you must use a more complete path as the first item in the `RESTART` keywod.

**Comment by loveduah1** [comment]
Thank you.

Please a second look at my tutorial plots show some variations in the WBHP and WGPR plots (comparing with the plots from the tutorial on the OPM site). How do I explain these differences? 
[Tutorial.zip](https://github.com/OPM/opm-simulators/files/8329315/Tutorial.zip)

**Comment by OPMUSER** [comment]
@loveduah1 I'm guessing the example on the web site is probably STALE, the real check should be against E100.  When I do this I get a perfect match as you can see below:

![Multi_Plot_1](https://user-images.githubusercontent.com/17958927/159624117-f036782c-5e9e-4441-962d-62887ed4fd5b.png)

In the above plot the symbols represent OPM Flow results and the lines E100.

Note you should install OPM ResInsight for plotting and 3D visualization. If you are on Windows machine install the Windows version, even if you are running OPM Flow under WSL. Otherwise install the Linux version.

![SPE1CASE1_3D_View_PRESSURE_00_01_Feb_2015](https://user-images.githubusercontent.com/17958927/159624859-def71007-f03d-449b-a69b-61a4615410b5.png)

Read the documentation for OPM Flow (on the web) and the OPM Flow Manual.

Now back to billable work :-)

**Comment by loveduah1** [comment]
Thank you

**Comment by thewillingone** [comment]
Hello there @loveduah1 , I see that you are also well versed with the OPM Flow even though you are also new. May I humbly seek your assistance on the issue that I posted about a relevant error I had encountered in my simulation. It's the most recent issue for the Error: Solver failed to converge after cutting timestep 10 times.  Your help would be highly appreciated.

**Comment by loveduah1** [comment]
@thewillingone the last time I saw an error like this was when there were discrepancies between the WCONINJE and my timesteps. I think cutting only the timesteps will not give this error.  OPM  successfully runs on only two timesteps for me at the moment. Usually submitting the input file [filename].DATA can help fix the error very easily.

**Comment by loveduah1** [comment]
> > Please `SKIPREST` keyword returns the error above
> 
> In this case it's not actually `SKIPREST`. Rather it's the `RESTART` keyword:
> 
> > ```
> > Error: Problem with keyword RESTART
> > In PROJ1.DATA line 255
> > The restart file: PROJ1.X0002 does not exist
> > ```
> 
> The `RESTART` keyword in your input file is
> 
> ```
> SOLUTION
> -- -------------------------------------------------------------------------
> --item #:      1        2         3          4    
> RESTART
>             PROJ1      2         1*         1*  /
> ```
> 
> Moreover, you've got `UNIFOUT` in your `RUNSPEC` section which means that the base run produces unified output files (`PROJ1.SMSPEC`, `PROJ1.UNSMRY`, and `PROJ.UNRST`).
> 
> Did you already produce a base run that you're trying to restart from? If not, the `RESTART` keyword is not appropriate here. If you **did** produce a base run, do you have the `PROJ1.UNRST` file somewhere? If you do, then please insert the keyword `UNIFIN` next to the `UNIFOUT` keyword in the `RUNSPEC` section. You will also use a more complete path if the `PROJ1.UNRST` file is in a different location than your `PROJ1.DATA` file.
> 
> Otherwise, if you have separate restart files named `PROJ1.X*` (e.g., `PROJ1.X0002`) somewhere that's not in the same directory as your `PROJ1.DATA`, then you must use a more complete path as the first item in the `RESTART` keywod.

Please the 'SKIPREST' with the 'RESTART' keywords seems to ignore any changes made in the new input file. ie., it skips to where the base run stopped without taking into consideration new changes in injection/production rates. Please how can I RESTART where I left off but continue with new injection/production conditions?

**Comment by bska** [comment]
> Please the `SKIPREST` with the `RESTART` keywords [...] skips to where the base run stopped without taking into consideration new changes in injection/production rates.

For what it's worth what you describe is the intended behaviour of `SKIPREST`.  In particular, the keyword lets you use the `SCHEDULE` section of an existing `.DATA` file and will `SKIP` all `SCHEDULE` information up to and including the `REST`art time.

> How can I RESTART where I left off but continue with new injection/production conditions?

If you want to impose new constraints in a restarted run you need to insert those `WCONINJE` and `WCONPROD` keywords **after** the restart time.  However, since you're using `TSTEP` I think it might be easier to remove SKIPREST altogether and just create the wells and groups from the restart file.  In other words, I'd use one `.DATA` file for the base run and then create a separate input file, e.g., `PROJ_RESTART.DATA` for the restarted run.  The easiest way to create this input file is to just copy the base run input file,
```sh
cp PROJ.DATA PROJ_RESTART.DATA
```
and then change the `SOLUTION` and `SCHEDULE` sections in `PROJ_RESTART.DATA` to look similar to
```
SOLUTION
RESTART
  PROJ 2 / -- <--- Or some other restart step ID here

SCHEDULE
-- -------------------------------------------------------------------------
RPTSCHED
	'PRES' 'SGAS' 'RS' 'WELLS' /

RPTRST
	'BASIC=2' /

WCONPROD
-- Item #:1	2      3     4	   5  9
	'PROD' 'OPEN' 'ORAT' 20000 4* 1000 /
/

WCONINJE
-- Item #:1	 2	 3	 4	5      6  7
	'INJ'	'GAS'	'OPEN'	'RATE'	10000 1* 9014 /
	'INJ2'  'GAS'	'OPEN'	'RATE'	10000 1* 9014 /	
/

TSTEP
--Advance the simulator once a month for TEN years:
31 28/

--Advance the simulator once a year for TEN years:
-- /
```
If you do it this way, then the simulator assumes that the all wells are stored in the restart file (`PROJ.UNRST`) and will just continue using those.  The results of the restarted simulation will be output to files named `PROJ_RESTART.SMSPEC`, `PROJ_RESTART.UNSMRY`, and `PROJ_RESTART.UNRST`.

**Comment by OPMUSER** [comment]
@bska You have the patience of a Saint.

**Comment by loveduah1** [comment]
> > Please the `SKIPREST` with the `RESTART` keywords [...] skips to where the base run stopped without taking into consideration new changes in injection/production rates.
> 
> For what it's worth what you describe is the intended behaviour of `SKIPREST`. In particular, the keyword lets you use the `SCHEDULE` section of an existing `.DATA` file and will `SKIP` all `SCHEDULE` information up to and including the `REST`art time.
> 
> > How can I RESTART where I left off but continue with new injection/production conditions?
> 
> If you want to impose new constraints in a restarted run you need to insert those `WCONINJE` and `WCONPROD` keywords **after** the restart time. However, since you're using `TSTEP` I think it might be easier to remove SKIPREST altogether and just create the wells and groups from the restart file. In other words, I'd use one `.DATA` file for the base run and then create a separate input file, e.g., `PROJ_RESTART.DATA` for the restarted run. The easiest way to create this input file is to just copy the base run input file,
> 
> ```shell
> cp PROJ.DATA PROJ_RESTART.DATA
> ```
> 
> and then change the `SOLUTION` and `SCHEDULE` sections in `PROJ_RESTART.DATA` to look similar to
> 
> ```
> SOLUTION
> RESTART
>   PROJ 2 / -- <--- Or some other restart step ID here
> 
> SCHEDULE
> -- -------------------------------------------------------------------------
> RPTSCHED
> 	'PRES' 'SGAS' 'RS' 'WELLS' /
> 
> RPTRST
> 	'BASIC=2' /
> 
> WCONPROD
> -- Item #:1	2      3     4	   5  9
> 	'PROD' 'OPEN' 'ORAT' 20000 4* 1000 /
> /
> 
> WCONINJE
> -- Item #:1	 2	 3	 4	5      6  7
> 	'INJ'	'GAS'	'OPEN'	'RATE'	10000 1* 9014 /
> 	'INJ2'  'GAS'	'OPEN'	'RATE'	10000 1* 9014 /	
> /
> 
> TSTEP
> --Advance the simulator once a month for TEN years:
> 31 28/
> 
> --Advance the simulator once a year for TEN years:
> -- /
> ```
> 
> If you do it this way, then the simulator assumes that the all wells are stored in the restart file (`PROJ.UNRST`) and will just continue using those. The results of the restarted simulation will be output to files named `PROJ_RESTART.SMSPEC`, `PROJ_RESTART.UNSMRY`, and `PROJ_RESTART.UNRST`.

thank you.

**Comment by loveduah1** [comment]
> > > Please the `SKIPREST` with the `RESTART` keywords [...] skips to where the base run stopped without taking into consideration new changes in injection/production rates.
> > 
> > 
> > For what it's worth what you describe is the intended behaviour of `SKIPREST`. In particular, the keyword lets you use the `SCHEDULE` section of an existing `.DATA` file and will `SKIP` all `SCHEDULE` information up to and including the `REST`art time.
> > > How can I RESTART where I left off but continue with new injection/production conditions?
> > 
> > 
> > If you want to impose new constraints in a restarted run you need to insert those `WCONINJE` and `WCONPROD` keywords **after** the restart time. However, since you're using `TSTEP` I think it might be easier to remove SKIPREST altogether and just create the wells and groups from the restart file. In other words, I'd use one `.DATA` file for the base run and then create a separate input file, e.g., `PROJ_RESTART.DATA` for the restarted run. The easiest way to create this input file is to just copy the base run input file,
> > ```shell
> > cp PROJ.DATA PROJ_RESTART.DATA
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and then change the `SOLUTION` and `SCHEDULE` sections in `PROJ_RESTART.DATA` to look similar to
> > ```
> > SOLUTION
> > RESTART
> >   PROJ 2 / -- <--- Or some other restart step ID here
> > 
> > SCHEDULE
> > -- -------------------------------------------------------------------------
> > RPTSCHED
> > 	'PRES' 'SGAS' 'RS' 'WELLS' /
> > 
> > RPTRST
> > 	'BASIC=2' /
> > 
> > WCONPROD
> > -- Item #:1	2      3     4	   5  9
> > 	'PROD' 'OPEN' 'ORAT' 20000 4* 1000 /
> > /
> > 
> > WCONINJE
> > -- Item #:1	 2	 3	 4	5      6  7
> > 	'INJ'	'GAS'	'OPEN'	'RATE'	10000 1* 9014 /
> > 	'INJ2'  'GAS'	'OPEN'	'RATE'	10000 1* 9014 /	
> > /
> > 
> > TSTEP
> > --Advance the simulator once a month for TEN years:
> > 31 28/
> > 
> > --Advance the simulator once a year for TEN years:
> > -- /
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
[attachments.zip](https://github.com/OPM/opm-simulators/files/8564952/attachments.zip)

> >   
> > If you do it this way, then the simulator assumes that the all wells are stored in the restart file (`PROJ.UNRST`) and will just continue using those. The results of the restarted simulation will be output to files named `PROJ_RESTART.SMSPEC`, `PROJ_RESTART.UNSMRY`, and `PROJ_RESTART.UNRST`.
> 
> thank you.

 I am getting the error when I restart the attached files. please what am I doing wrong?


Error: 
An error occurred while creating the reservoir properties
Internal error: Tried to get uninitialized value from DeckItem index: 0


Error: Unrecoverable errors were encountered while loading input

**Comment by loveduah1** [comment]
-Please I tried to restart from the attached files. I made all the changes
you recommended yet I still can control the injectors outside the base
file.

see attached

**Comment by OPMUSER** [comment]
You were missing the SOLUTION section to load the RESTART file - see the attached.

[SKIPREST-RESTART.zip](https://github.com/OPM/opm-simulators/files/8569565/SKIPREST-RESTART.zip)

**Comment by bska** [comment]
> I am getting the error when I restart the attached files
>
> ```
> Error: An error occurred while creating the reservoir properties Internal error: Tried to get uninitialized value from DeckItem index: 0
> 
> Error: Unrecoverable errors were encountered while loading input
> ```

First, unless something went wrong in packaging up the "attachments.zip" file, the `RESTART` keyword of `PROJ_RESTART.DATA` is entered as
```
RESTART
PROJ2 /
```
with no separation between the baserun name `PROJ` and the report step ID `2`.  You need to write
```
RESTART
PROJ 2 / -- <-- Note space between 'PROJ' and '2' here
```
instead.  Second, the `RESTART` keyword should be the **only** data in the `SOLUTION` section of a restarted run.  Flow happens to accept `EQUIL` and other keywords here, but you really should remove all that information in a restarted run.  Third, you still have `WELSPECS` and `COMPDAT` in your `PROJ_RESTART.DATA` file.  Unless you're introducing new wells (or new reservoir connections) in a restarted run those should not be in the `SCHEDULE` section.  Finally, you've not removed the `SKIPREST` keyword as I said you should.  That is the fundamental reason we ignore your `WCONPROD` and `WCONINJE` setting.  I really did mean it when I [gave the example](https://github.com/OPM/opm-simulators/issues/3846#issuecomment-1085715998) of a `SCHEDULE` section that had neither `WELSPECS`, `COMPDAT`, nor `SKIPREST`.

That said though, I'm not convinced that simulation restart is the appropriate tool for your workflow.  Simulation restart is targeted at the following use case

* Base run simulates the time period from T=start&mdash;in your case 1st December 2014&mdash;to T=T1
* Restarted run simulates a time period starting at T=T1 onward to T=T2 with T2>T1

One common application of this setup in a reservoir engineering workflow is to simulate the historical period of a field once in the base run and then to simulate multiple future development and management scenarios in restarted runs that all use the end result of the base run as their initial condition.  I got the impression that your reinforcement learning workflow wanted to simulate the **same** time period from `START` to end in all runs and in that case simulation restart as represented by the `RESTART` keyword does not really give any useful aid.  On the other hand if the reinforcement learning can be meaningfully separated into two (or more) time periods where the end result of one period is the initial condition for the next then simulation restart is one way to go about it.

**Comment by loveduah1** [comment]
Please I am on it

**Comment by OPMUSER** [comment]
@loveduah1 did the example I sent not work on your machine?

**Comment by loveduah1** [comment]
Please I will look at it when I go to school today.

**Comment by loveduah1** [comment]
> You were missing the SOLUTION section to load the RESTART file - see the attached.
> 
> [SKIPREST-RESTART.zip](https://github.com/OPM/opm-simulators/files/8569565/SKIPREST-RESTART.zip)

Error: Problem with keyword DATES
In PROJ1_RESTART.DATA line 403
In a restarted simulation using SKIPREST, the DATES keyword must have
a record corresponding to the RESTART time 01-Feb-2015 00:00:00.
Reached time 01-Apr-2015 00:00:00 without an intervening record.

Error: Unrecoverable errors were encountered while loading input




please I tried to make changes to your file to see the results. I keep getting the error above. please is it restarting from FEB 01.  I'm not sure where to change the restart time.

[SKIP.zip](https://github.com/OPM/opm-simulators/files/8574928/SKIP.zip)

**Comment by bska** [comment]
> I tried to make changes to your file to see the results. I keep getting the error above. please why is it restarting from FEB 01.

There are, fundamentally, two different strategies to specifying a restarted simulation run.

1. Keep all existing `SCHEDULE` data **and** insert `SKIPREST` at the beginning of the `SCHEDULE` section.  This mode typically uses the `DATES` keyword to advance the simulator from one point in time to the next.  This mode requires a `DATES` record corresponding to the restart time in order for the restart mechanism to properly locate where to start simulating.  For what it's worth he diagnostic message
    > ```
    > In a restarted simulation using SKIPREST, the DATES keyword must have
    > a record corresponding to the RESTART time 01-Feb-2015 00:00:00.
    > Reached time 01-Apr-2015 00:00:00 without an intervening record.
    > ```
    is trying to tell you that there is no `DATES` record corresponding to the restart time (1st February 2015) and that you therefore must insert such a record&mdash;at an appropriate location, possibly after your `WCONINJE` keyword on lines 396&ndash;400&mdash;to assist the restart mechanism.

2. Remove all `SCHEDULE` data up to **and including** the restart time.  This mode, therefore, assumes that the restart time is at the start of the `SCHEDULE` section so there should typically **not** be a `DATES` record corresponding to the restart time in this case.  You will typically also use the `TSTEP` keyword to advance the simulator from one point in time to the next, but `DATES` is supported too.

@OPMUSER's example is set up using strategy 1 while I suggested opting for strategy 2.  Either strategy works if you honour the restrictions and there are benefits and drawbacks to both of them.

Regardless of which strategy you use do keep in mind that restarted simulations are structured in the way I [outlined earlier](https://github.com/OPM/opm-simulators/issues/3846#issuecomment-1110745739) with the base run simulating from the `START` time until some later time `T1` and then the restarted simulation starting at `T1` and running until some later time `T2`.  If this behaviour is at odds with what you're trying to do then restarted simulations may not be the correct tool for your use case.

> I'm not sure where to change the restart time.

The restart time is directly tied to the restart step ID that you supply to the `RESTART` keyword.  Your `PROJ1_RESTART` input file has
```
RESTART
  PROJ1 1 1* 1*/
```
and the value `1` supplied in item 2 here means that the restart time is at the end of the first report step of the `PROJ1` base run.  If you specify a different report step in the second item then you implicitly change the restart time too.  You can get a list of available restart steps and their associate restart times stored in the `PROJ1.UNRST` file by using the `-l` (ell) option of the `convertECL` utility:
```sh
convertECL -l PROJ1.UNRST
```

**Comment by loveduah1** [comment]
Thank you very much, Sir.

**Comment by loveduah1** [comment]
Error: Record contains too many items in keyword RESTART. Expected 4 items, found 5.
In file PROJ1_RESTART.DATA at line 342.
Record is "'PROJ1'                1         2*        1*     ".

Error: Unrecoverable errors were encountered while loading input
phd@elfd:~/OPMPROJ/PHDNEW/SKIPREST/try$ convertECL -l PROJ1.UNRST
Report step number:    1   Date: 2015/02/01
Report step number:    2   Date: 2015/03/01
Report step number:    3   Date: 2015/04/01

Please this is the error I am getting

**Comment by loveduah1** [comment]
please I am sorry I corrected the wrong item

**Comment by loveduah1** [comment]
phd@elfd:~/OPMPROJ/PHDNEW/SKIPREST/try$ convertECL -l PROJ1.UNRST
Report step number:    1   Date: 2015/02/01
Report step number:    2   Date: 2015/03/01
Report step number:    3   Date: 2015/04/01
phd@elfd:~/OPMPROJ/PHDNEW/SKIPREST/try$ convertECL -l PROJ1_RESTART.UNRST
terminate called after throwing an instance of 'std::runtime_error'
  what():  Can not open EclFile: PROJ1_RESTART.UNRST
Aborted (core dumped)

please how can I see the new timestep in the PROJ1_RESTART.UNRST

Also I dont see anything new new timesteps on ResInsight.

**Comment by OPMUSER** [comment]
> please how can I see the new timestep in the PROJ1_RESTART.UNRST
> Also I don't see any new timesteps on ResInsight.

See attached.
[PROJ1_RESTART2.zip](https://github.com/OPM/opm-simulators/files/8579189/PROJ1_RESTART2.zip)

**Comment by loveduah1** [comment]
Thank you.

**Comment by OPMUSER** [comment]
@loveduah1 Some people are never happy :-)

This is a test model and running in seconds is very good when compared to actual field models that take hours and in some cases days to run. If the intent is to apply this methodology to an actual field model then then take at look at the Norme model https://github.com/OPM/opm-tests/tree/master/norne 

To improve run time you can throw more hardware at the model, so more CPUs and threads using mpi. For example:

`mpirun -np 4 flow -CASENAME.DATA`

Will use four cores to run CASNAME.DATA. 

Tuning models to run faster is not a simple task, so I think the above is your best option for the moment.

**Comment by bska** [comment]
> it takes approx. 1.56s to complete one episode (120 timesteps). For the algorithm to efficiently learn, I need at least 1000-50,000 episodes. So I am hoping to find a way for the algorithm to skip through timesteps.

It's not clear to me what you mean by "skip through timesteps" here.  The model runs in a few seconds and a sizeable fraction of that will be to read the input file and to write the output file.  That is inherently a sequential process and places limits on the achievable speedup from going to a parallel execution (`mpirun` as suggested by @OPMUSER).

Anyway, do the simulation/learning results of episode E depend on the simulation/learning results of episode E-1 or any of the earlier ones?  If not, you could probably try to run multiple episodes concurrently.  On the other hand if the learning results **do** depend on having learning results from earlier episodes available then there's not really a lot you can do.  The episodes must be run one after the other in this case.

**Comment by loveduah1** [comment]
Please how do I format one INJ to pump either WATER or GAS?
WELLSPEC section only allows one.
WELSPECS
-- Item #: 1	 2	3	4	5	 6
	'PROD'	'G1'	6	1	8400	'OIL' /
	'INJ1'	'G1'	1	1	8335	'GAS' /

I want  to be able to alternate GAS and WATER like this
 
DEC 2014/
/
WCONINJE
'INJ1' WATER OPEN RATE 10000.0/
/
 
DATES
1 JAN 2015 /
/
WCONINJE
'INJ1' GAS OPEN RATE 10000.0/

**Comment by OPMUSER** [comment]
@loveduah1 Have you tried this as it should work? `WELSPECS` sets the preferred phase, whereas `WCONINJE` sets the injection fluid and rates.

**Comment by loveduah1** [comment]
Please I will try it now. Thank you

**Comment by loveduah1** [comment]
>>On the other hand if the reinforcement learning can be meaningfully separated into two (or more) time periods where the end result of one period is the initial condition for the next then simulation restart is one way to go about it.

please I am getting this error but not sure why PROJ1_RESTART.UNRST is not generating for the third restart in the series


creating grid from keywords DX, DY, DZ and TOPS

Error: Problem with keyword RESTART
in PROJ1_RESTART2.DATA line 342
The restart file: PROJ1_RESTART1.UNRST does not exist

Error: Unrecoverable errors were encountered while loading input
[SKIPREST.zip](https://github.com/OPM/opm-simulators/files/8655483/SKIPREST.zip)

**Comment by OPMUSER** [comment]
@loveduah1 There are many errors in the input decks, including:

- Multiple SOLUTION and SUMMARY sections.
- Outputting the RESTART file as unified but requesting the restart run using non-unified.
- Miss-match of dates, the RESTART record date, say 1 FEB 2015 must be in the RESTART file and the SCHEDULE section so the simulator knows how to match the two.
- You should not change the SCHEDULE section before the restart date, only after.
- You have to have SCHEDULE keywords after the restart date, including at least one more DATE keyword.
- Changing the START date in the RUNSPEC section should not be done.
- Using BASIC=4 in the RPTRST keyword in the restart runs instead of BASIC=1. The latter will write out a restart record every report step but keep only the last one.

Having said all this, I get `PROJ2` and` PROJ2-RESTART1` cases to run as expected; however,  `PROJ2-RESTART2` fails and so will  therefore `PROJ2-RESTART3`. This failure is a bug.

@bska  `PROJ2-RESTART2` fails with:

`Error: Unrecoverable errors were encountered while loading input`

I thought this was may be related to which restart record to restart from. So I tried both 1 and 2, but both fail. The commercial simulator restart records are cumulative so it should be record number 2.

Test deck and results attached.

[PROJ2.zip](https://github.com/OPM/opm-simulators/files/8657464/PROJ2.zip)
[PROJ2_RESTART1.zip](https://github.com/OPM/opm-simulators/files/8657465/PROJ2_RESTART1.zip)
[PROJ2_RESTART2.zip](https://github.com/OPM/opm-simulators/files/8657468/PROJ2_RESTART2.zip)
[PROJ2_RESTART3.zip](https://github.com/OPM/opm-simulators/files/8657469/PROJ2_RESTART3.zip)

All cases run as expected with the commercial simulator.

**Comment by loveduah1** [comment]
> @loveduah1 Have you tried this as it should work? `WELSPECS` sets the preferred phase, whereas `WCONINJE` sets the injection fluid and rates



WELSPECS
-- Item #: 1	 2	3	4	5	 6
	'PROD'	'G1'	6	1	8400	'OIL' /
	'INJ1'	'G1'	1	1	8335	'GAS' /


It works, thank you.
Please what 'GAS' is being injected? Is it CO2? how can I inject specifically CO2?

**Comment by loveduah1** [comment]
> > `PROJ2-RESTART2` fails with:
> > ```
> > Error: Unrecoverable errors were encountered while loading input
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I thought this was may be related to which restart record to restart from. So I tried both 1 and 2, but both fail. The commercial simulator restart records are cumulative so it should be record number 2.
> > Test deck and results attached.
> 
> Thanks a lot for the test data. The poor diagnostic message is decidedly a bug in Flow, but the real issue is in the input `.DATA` files of `PROJ2_RESTART2.DATA` and `PROJ2_RESTART3.DATA`. Both of these have `DATES` keywords that are not properly terminated. In particular, `PROJ2_RESTART2.DATA` ends as follows
> 
> ```
> -- ...
> DATES
> 1 APR 2015
> /
> ECHO
> END
> ```
> 
> while `PROJ2_RESTART3.DATA` similarly ends as
> 
> ```
> -- ...
> DATES
> 1 MAY 2015
> /
> ECHO
> END
> ```
> 
> The commercial simulator is apparently more accepting than Flow's parser here and will infer that the `E` in `ECHO` cannot possibly be an integer and therefore must be the start of a new keyword. Flow, on the other hand, is rather strict in this respect and requires that all `DATES` records **and** the keyword itself be terminated by a `/` character.
> 
> In other words, ending `PROJ2_RESTART2.DATA` as
> 
> ```
> -- ...
> DATES
> 1 APR 2015 / -- <--- Note terminating '/' here
> /
> ECHO
> END
> ```
> 
> and `PROJ2_RESTART3.DATA` similarly as
> 
> ```
> -- ...
> DATES
> 1 MAY 2015 / -- <--- Note terminating '/' here
> /
> ECHO
> END
> ```
> 
> will allow Flow to simulate these two consecutive restart cases.

Thank you. I am on it

**Comment by OPMUSER** [comment]
> 
> Thanks a lot for the test data. The poor diagnostic message is decidedly a bug in Flow, but the real issue is in the input `.DATA` files of `PROJ2_RESTART2.DATA` and `PROJ2_RESTART3.DATA`. Both of these have `DATES` keywords that are not properly terminated. In particular, `PROJ2_RESTART2.DATA` ends as follows
> 
> ```
> -- ...
> DATES
> 1 APR 2015
> /
> ECHO
> END
> ```
@bska Apoligies I should have spotted that, but like you said the error message is distracting, good you are fixing that.

**Comment by OPMUSER** [comment]
> Please what 'GAS' is being injected? Is it CO2? how can I inject specifically CO2?

@loveduah1 Okay, first question what is the purpose of injecting CO2?

OPM Flow is a "black-oil" simulator, that is there is only oil, water and gas phases with limited mass transfer between the phases in the standard formulation. A compositional simulator models the various components, methane, ethane, CO2, etc. in the three phases and the mass transfer of the components to the various phases.

For "black-oil" models the Pressure Volume Temperature ("PVT") behavior is given via the PVT set of keywords, and for gas that is the PVDG keyword in your example. So if you derive the PVDG based on CO2 then the PVT behavior would model the injection of CO2.  However, injecting CO2 into an oil reservoir is an Enhanced Oil Recovery ("EOR") process as the CO2 is miscible with the oil, that is the CO2 mixes with the oil to swell the oil, which leads to increased oil recovery. So to correctly model CO2 injection you have to model the mass transfer of the CO2 into the oil. 

OPM Flow's Solvent model can do this. Take a look at https://github.com/OPM/opm-tests/tree/master/spe5_co2eor example to see the difference in the PVT keywords. So what you need to do, because you don't have access to PVT Equation of State("EOS") software, is to use the PVT in the aforementioned example, with your data set. That is you need to replace your PVT data with the example PVT data. Fortunately, they are both in Field units.

Take your time in doing this to avoid errors.

**Comment by loveduah1** [comment]
Thank you.

**Comment by loveduah1** [comment]
@loveduah1  Okay, first question what is the purpose of injecting CO2?

@OPMUSER  
Sorry, I just saw this.

I want to start incorporating CO2 into my setup now because my main project is about co2 sequestration.

**Comment by OPMUSER** [comment]
@loveduah1 Well if it is CO2 sequestration then that is a different matter. What I described above is how to model a CO2 EOR oil project, modelling CO2 sequestration is different. 

Take a look at Sandve et al.  excellent paper on this subject:

_Tor Harald Sandve1, Sarah E. Gasda, Atgeirr Rasmussen, and Alf Birger Rustad. Convective dissolution in field scale CO2 storage simulation using the OPM Flow simulator. Submitted to TCCS 11 – Trondheim Conference on CO2 Capture, Transport and Storage Trondheim, Norway – June 21-23, 2021._

The paper basically describes the workflow and the what has been implemented in OPM Flow. Also read the OPM Flow manual, in particular the CO2STORE, DIFFUSE, THERMAL, RTEMPVD  and SALINITY keywords. Then take a look at the examples: https://github.com/OPM/opm-tests/tree/master/co2store.

The actual deck will depend on the target reservoir:

1. a depleted oil field/reservoir,
2. a depleted gas field/reservoir, or
3. a saline aquifer?

Fortunately, you don't have to derive the CO2 PVT data as the authors of the paper incorporated these in the the simulator.

**Comment by loveduah1** [comment]
please saline aquifer.

**Comment by loveduah1** [comment]
> > Please what 'GAS' is being injected? Is it CO2? how can I inject specifically CO2?
> 
> @loveduah1 Okay, first question what is the purpose of injecting CO2?
> 
> OPM Flow is a "black-oil" simulator, that is there is only oil, water and gas phases with limited mass transfer between the phases in the standard formulation. A compositional simulator models the various components, methane, ethane, CO2, etc. in the three phases and the mass transfer of the components to the various phases.
> 
> For "black-oil" models the Pressure Volume Temperature ("PVT") behavior is given via the PVT set of keywords, and for gas that is the PVDG keyword in your example. So if you derive the PVDG based on CO2 then the PVT behavior would model the injection of CO2. However, injecting CO2 into an oil reservoir is an Enhanced Oil Recovery ("EOR") process as the CO2 is miscible with the oil, that is the CO2 mixes with the oil to swell the oil, which leads to increased oil recovery. So to correctly model CO2 injection you have to model the mass transfer of the CO2 into the oil.
> 
> OPM Flow's Solvent model can do this. Take a look at https://github.com/OPM/opm-tests/tree/master/spe5_co2eor example to see the difference in the PVT keywords. So what you need to do, because you don't have access to PVT Equation of State("EOS") software, is to use the PVT in the aforementioned example, with your data set. That is you need to replace your PVT data with the example PVT data. Fortunately, they are both in Field units.
> 
> Take your time in doing this to avoid errors.



@OPMUSER 
Please if I want to increase oil production in the tertiary production stage but with the added advantage of storing CO2 in the depleted reservoir will that be CO2EOR?

**Comment by OPMUSER** [comment]
Yes.

OPM User

**Comment by OPMUSER** [comment]
I get this to run fine without modification using the latest release 2022-04

![image](https://user-images.githubusercontent.com/17958927/171323583-d12711e0-02f5-4f01-a33f-cee447786866.png)

![image](https://user-images.githubusercontent.com/17958927/171323681-7e2203fc-7963-420c-a5cb-fe92508e3f3a.png)

[Uploading EORPROJ.zip…]()

For the other job, EORPVT, version, you have extra `"\"`

![image](https://user-images.githubusercontent.com/17958927/171325878-420f2985-06a1-4209-91c9-1cf83afd526d.png)

And the Bg is not monotonically decreasing, if I comment the last value out as shown below:

![image](https://user-images.githubusercontent.com/17958927/171327301-fd2f8d0f-ac1d-4d5e-9a5a-b675310df806.png)

the job runs -see attached.

[EORPVT-01.zip](https://github.com/OPM/opm-simulators/files/8811034/EORPVT-01.zip)

**Comment by loveduah1** [comment]
**SCHEDULE ONE**
WCONINJE
-- Item #:1	 2	 3	 4	5      6   7
'INJW'	'WATER'	'OPEN'	'RATE'	12000 1* 10000 /
'INJG'	'GAS'	'OPEN'	'RATE'	12000 1* 10000 /
/
WSOLVENT
'INJG'  1.0/
/
WELOPEN
 'INJW' 'SHUT' /
 'INJG' 'SHUT' /
/
-- Start production for two months
TSTEP
--Advance the simulator once a month for two months:
31 28 
/
-- Start WAG injection after 2 months 
WELOPEN
 'INJW' 'OPEN' /
 'INJG' 'SHUT' /
/
TSTEP
--Advance the simulator once a month for ONE month:
31 
/


**SCHEDULE TWO**
WCONINJE
-- Item #:1	 2	 3	 4	5      6   7
	'INJW'	'WATER'	'OPEN'	'RATE'	12000 1* 10000 /
	'INJG'	'GAS'	'OPEN'	'RATE'	12000 1* 10000 /
/
WSOLVENT
'INJG'  1.0/
/

DATES
1 FEB 2014 /
/
WCONINJE
'INJW' GAS SHUT RATE 10000.0/
'INJG' GAS SHUT RATE 10000.0/
/
DATES
1 MAR 2014 /
/
WCONINJE
'INJW' GAS SHUT RATE 10000.0/
'INJG' GAS SHUT RATE 10000.0/
/
DATES
1 APR 2014 /
/
WCONINJE
'INJW' GAS OPEN RATE 10000.0/
'INJG' GAS SHUT RATE 10000.0/
/
 

Please I wanted to format Schedule One so that I can have control over both the WAG injection and production periods so I for it as the Schedule Two. please the main purpose of this is to know how long WAG injection takes.

Although FOPR is the same for both settings, BHP for the schedules is different for the injectors (attached). How can I fix this?
[EORWAG.zip](https://github.com/OPM/opm-simulators/files/8815162/EORWAG.zip)

**Comment by loveduah1** [comment]
Error: Problem with keyword RESTART
In WAG_RESTART.DATA line 318
The restart file: WAG.X0001 does not exist

Error: Unrecoverable errors were encountered while loading input


Please I am getting the above error during restart. I expected to restart from the FILE.UNRST file but it did not. How can I restart from the UNRST file?

[WAGSKIP.zip](https://github.com/OPM/opm-simulators/files/8816061/WAGSKIP.zip)

**Comment by loveduah1** [comment]
Thank you

**Comment by loveduah1** [comment]
Error: Record contains too many items in keyword GRAVITY. Expected 3 items, found 10.
In file MINES.DATA at line 49.
Record is "1   1   36
UNIFIN
UNIFOUT
GRID
INIT
NOECHO
DX
400*1000 ".
[MINES.zip](https://github.com/OPM/opm-simulators/files/8879514/MINES.zip)
Error: Unrecoverable errors were encountered while loading input
Please I activated some Keywords using the information attached but I keep getting errors above. 
Please how do I fix this?

**Comment by bska** [comment]
> ```
> Error: Record contains too many items in keyword GRAVITY. Expected 3 items, found 10.
> In file MINES.DATA at line 49.
> Record is "1   1   36
> UNIFIN
> ...
> ```

I suspect that you get into this situation because Flow's input parser is more strict than ECLIPSE's in this particular situation.  Your `MINES.DATA` input file is as follows starting from line 49, i.e., the line number identified in the diagnostic message
```
GRAVITY
      1   1   36
--       SWITCH ON THE UNIFIED INPUT FILES OPTION
--
UNIFIN
```
Flow requires that each record of the `GRAVITY` keyword be properly terminated by a slash character (`/`) whereas ECLIPSE appears to be able to infer that the keyword `UNIFIN` is not actually part of the `GRAVTIY` data.

Please insert a slash character after the value `36` on line 50 to make the `GRAVITY` data look like
```
GRAVITY
      1   1   36 /
--       SWITCH ON THE UNIFIED INPUT FILES OPTION
--
UNIFIN
```
and then try to rerun your case.  Do note, however, that these kinds of heavy oils are rarely tested so you may run into difficulties getting the simulation to converge.

**Comment by loveduah1** [comment]
Thank you. Please it worked however I got this error now.

Error: Too many records in keyword PVDG
In PVDG line MINES.DATA

Error: Unrecoverable errors were encountered while loading input


Also, is there a way to remove the constraint on item 5  in WCONPROD or how do I set a target that is impossible to achieve? I need this target above average so that no two episodes can receive similar returns (FOPT) during training.

WCONPROD
-- Item #:1	2      3     4	   5  9
	'PROD' 'OPEN' 'ORAT' 100 4* 2000 /
[MINES.zip](https://github.com/OPM/opm-simulators/files/8881019/MINES.zip)

**Comment by loveduah1** [comment]
Processing keyword WCONINJE at line 330

Processing grid
Total number of active cells: 400 / total pore volume: 419443414 RB

===============Saturation Functions Diagnostics===============

System:  Black-oil system.
Relative permeability input format: Saturation Family I.
Number of saturation regions: 1

Program threw an exception: EclProblem::_updateCompositionLayers() failed: Requested the internal energy of oil but it is disabled
[MINES.zip](https://github.com/OPM/opm-simulators/files/8882959/MINES.zip)


Please why am I getting this error code?

**Comment by OPMUSER** [comment]
Okay, you keep on placing the keywords in the wrong section, for example the `GRAVITY` and `RTEMP` keywords in your deck are in the `RUNSPEC` section when they should be in the `PROPS` section. Also, the values on the `GRAVITY` keyword seem unreasonable gas gravity of 36 (air=1) for example - this will cause numeric problems. Also, you either enter the GRAVITY or the DENSITY keywords, but not both.

You used the THERMAL keyword in the RUNSPEC section which means that you have to enter the `THCONR`, `SPECHEAT` and  `SPECROCK`, etc.  keywords, if this is indeed a thermal run, and this is the reason for the error. In the attached deck I have switched the THERMAL option off.

> Also, is there a way to remove the constraint on item 5 in WCONPROD or how do I set a target that is impossible to achieve? I need this target above average so that no two episodes can receive similar returns (FOPT) during training.

In your model you have to set a well target, otherwise  the simulator does not know how to produce the well, so what you hae now is:
```
--
--       WELL PRODUCTION WELL CONTROLS                                                     
--                                                                              
-- WELL  OPEN/  CNTL   OIL    WAT    GAS   LIQ    RES    BHP   THP   VFP    VFP  
-- NAME  SHUT   MODE   RATE   RATE   RATE  RATE   RATE   PRES  PRES  TABLE  ALFQ 
WCONPROD                                                                    
'PROD'  'OPEN' 'ORAT'  100    1*     1*    1*     1*     2000                    /
/

```
So the target is 100 stb/d with a minimum bottom-hole pressure ("BHP") constraint of 2000 psia (which seems very high).  So you could just change the 100 stb/d target to an unrealistc higher value and use a sensible BHP of between 200 to 500 psia.  You need the BHP constraint as this is a physical limit for the well. Normally Vertical Lift Performance ("VFP") tables are used in a model to convert from the simulators BHP to the well head, that is the tubing-head pressure ("THP"). Its is the THP, that is wellhead pressure at the surface, that is the correct physical constraint for a producing well.

**Comment by OPMUSER** [comment]
Sorry forgot the attachment.

[MINES-01.zip](https://github.com/OPM/opm-simulators/files/8883310/MINES-01.zip)

**Comment by loveduah1** [comment]
Thank you, Sir.

**Comment by loveduah1** [comment]
Please how can I get the oil recovery factor to generate the attached plots? 
I added 'FOE'  to the summary section but there were no results in the file.RSM.
[ORF.zip](https://github.com/OPM/opm-simulators/files/8892691/ORF.zip)

**Comment by bska** [comment]
> I added 'FOE' to the summary section but there were no results in the `.RSM` file.

Alas, the simulator does not currently output `FOE`.  PR OPM/opm-common#3049 rectifies this problem, but the fix won't be available in a public binary until the 2022.10 release later this year. Unless you build the simulator from source code you will not get that output directly.  In that case you'll have to approximate the recovery factor using the `FOIP` summary vector, e.g., something like
```
FOE \approx (FOIP(1) - FOIP) / FOIP(1)
```
with `FOIP(1)` denoting the initial/first value of `FOIP`, which should be close to the initial oil in place, and `FOIP` denoting the current oil in place.

**Comment by loveduah1** [comment]
Thank you

**Comment by loveduah1** [comment]
I could output FOIP in the RSM file but not the solution to "FOE \approx (FOIP(1) - FOIP) / FOIP(1)'

Do I have to manually calculate the results from the RSM data? How do I reference the variables so that I can output the approximated values?
[PROJ.zip](https://github.com/OPM/opm-simulators/files/8894058/PROJ.zip)

**Comment by bska** [comment]
> if you want to be clever, you can do this using the the UDQ keyword at the start of the SCHEDULE section:

True, with the additional provision that user-defined quantities at field level must have an initial name prefix of `FU`.  In other words, the UDQ definition will do the right thing with a slight renaming of the variables
```
--
-- DEFINE START OF USER DEFINED QUANTITY SECTION       
--     
UDQ
--
-- OPERATOR VARIABLE  EXPRESSION
--        
ASSIGN      FUOIIP   99.99            /
ASSIGN      FUFOE    0.0              /
DEFINE      FUFOE    (FOPT/FUOIIP )   /
UPDATE      FUFOE    ON               /

/  DEFINE END OF USER DEFINED QUANTITY SECTION
```

**Comment by OPMUSER** [comment]
@bska yep your right I forgot about that, see what happens when you don't test your code :-)

@loveduah1. Yes, that is correct except the value in the RSM file is in Mstb (thousands of stb) 
```
FOIP
STB
*10**3
```

so the correct value is 223287.5**E3**

**Comment by loveduah1** [comment]
Thank you

**Comment by bska** [comment]
I'm going to close this issue now since there's been no new comments for a month.  Please open a separate issue if something else comes up.
